File: examples/java-gradle/src/main/java/com/example/interfaces/Repository.java
package com.example.interfaces;

import java.util.List;
import java.util.Optional;

/**
 * Generic repository interface for CRUD operations.
 *
 * @param <T> the entity type
 */
public interface Repository<T> {

    /**
     * Find entity by ID.
     *
     * @param id the entity ID
     * @return optional containing the entity if found
     */
    Optional<T> findById(Long id);

    /**
     * Find all entities.
     *
     * @return list of all entities
     */
    List<T> findAll();

    /**
     * Save entity.
     *
     * @param entity the entity to save
     * @return the saved entity with ID
     */
    T save(T entity);

    /**
     * Delete entity by ID.
     *
     * @param id the entity ID
     * @return true if deleted
     */
    boolean delete(Long id);

    /**
     * Check if entity exists.
     *
     * @param id the entity ID
     * @return true if exists
     */
    default boolean exists(Long id) {
        return findById(id).isPresent();
    }

    /**
     * Count all entities.
     *
     * @return the count
     */
    default long count() {
        return findAll().size();
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/interfaces/Repository.kt
package com.example.interfaces

/**
 * Generic repository interface for CRUD operations.
 */
interface Repository<T> {

    /**
     * Find entity by ID.
     */
    fun findById(id: Long): T?

    /**
     * Find all entities.
     */
    fun findAll(): List<T>

    /**
     * Save entity.
     */
    fun save(entity: T): T

    /**
     * Delete entity by ID.
     */
    fun delete(id: Long): Boolean

    /**
     * Check if entity exists.
     */
    fun exists(id: Long): Boolean = findById(id) != null

    /**
     * Count all entities.
     */
    fun count(): Long = findAll().size.toLong()
}
